### PR TITLE
Replace MemoryManager implementation with rpmalloc

### LINKED ISSUE #3865: Poor Memory Manager performance
`MemoryManager` causes a performance regression and should be replaced by a faster algorithm or, if we fail to find/implement one, just `malloc` and `free`, which still provides better performance than the current implementation.

Here are a couple of existing lock-free allocators I found:

* ~~[Streamflow](http://people.cs.vt.edu/~scschnei/streamflow/) ([GitHub project](https://github.com/scotts/streamflow/)).~~ **Edit:** Implementation is POSIX-only.
  In case anyone's interested in the technical details, there's a paper describing the algorithm, including benchmarks: http://people.cs.vt.edu/~scschnei/papers/ismm06.pdf 
* [Lockless](http://locklessinc.com/)
* [TCMalloc](http://goog-perftools.sourceforge.net/doc/tcmalloc.html)
* [fc_malloc](https://github.com/bytemaster/fc_malloc)
* [rpmalloc](https://github.com/rampantpixels/rpmalloc)

Related issues: #2029, #3312, #3767

### PR DESCRIPTION
**Pull request by lukas-w**
Fixes #3865, by integrating [rpmalloc](https://github.com/rampantpixels/rpmalloc).

Quick – probably not very meaningful – benchmark, using Skiessi-RandomProjectNumber14253:

|   | time user  |  time system | elapsed  |  cpu |
|:---|---:|---:|---:|---:|
| stable-1.2  | 23.26  |  1.33 |  8.09 | 303%  |
| malloc/free in MemoryManager | 21.30  | 1.30  | 6.57  | 344%  |
| rpmalloc [override](https://github.com/rampantpixels/rpmalloc/blob/master/rpmalloc/malloc.c), replaces all memory allocations  | 21.66  | 1.45  | 6.47  | 357%  |
| rpmalloc (this PR)  | 21.15  | 1.45  | 6.71  | 337%  |

### CODE REVIEW

**Comment by zonkmachine** [comment]
Cool!

On Linuxmint 17.3, I get:

```
-- Configuring done
CMake Error at src/3rdparty/rpmalloc/CMakeLists.txt:3 (add_library):
  Cannot find source file:

    rpmalloc/rpmalloc/rpmalloc.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx

```

**Comment by tresf** [comment]
> On Linuxmint 17.3, I get:

It's a submodule, you'll have to clone with the recurse option.  `git clone --recursive foo://bar`

**Comment by Umcaruje** [comment]
Error when compiling with carla:
```
In file included from /home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp:25:0:
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.h:81:5: error: ‘QMutex’ does not name a type
     QMutex fMutex;
     ^
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp: In member function ‘intptr_t CarlaInstrument::handleDispatcher(NativeHostDispatcherOpcode, int32_t, intptr_t, void*, float)’:
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp:253:12: warning: enumeration value ‘NATIVE_HOST_OPCODE_INTERNAL_PLUGIN’ not handled in switch [-Wswitch]
     switch (opcode)
            ^
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp: In member function ‘virtual void CarlaInstrument::play(sample_t (*)[2])’:
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp:355:32: error: ‘fMutex’ was not declared in this scope
         const QMutexLocker ml(&fMutex);
                                ^
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp: In member function ‘virtual bool CarlaInstrument::handleMidiEvent(const MidiEvent&, const MidiTime&, f_cnt_t)’:
/home/umcaruje/github/lmmstest/plugins/carlabase/carla.cpp:371:28: error: ‘fMutex’ was not declared in this scope
     const QMutexLocker ml(&fMutex);
                            ^
plugins/carlabase/CMakeFiles/carlabase.dir/build.make:72: recipe for target 'plugins/carlabase/CMakeFiles/carlabase.dir/carla.cpp.o' failed
make[2]: *** [plugins/carlabase/CMakeFiles/carlabase.dir/carla.cpp.o] Error 1
```

**Comment by tresf** [comment]
@Umcaruje that appears to be a bug with Carla, it should have imported `QMutex` without relying on it from another header.  Can you add `#include <QMutex>` to the Carla plugin, it sounds like an existing bug with stable.

**Comment by Umcaruje** [comment]
Built without carla, I can confirm I don't get stutters while playing on elementary os loki under both qt4 and qt5. Very good and comparable performance with 1.1. I'll keep using it for further testing.

**Comment by PhysSong** [comment]
Tested. It showed very fast allocation speed and I got no crashes during tests.

File: src/core/MemoryManager.cpp [hunk 1]
```diff
@@ -25,197 +24,54 @@
...
-
-void MemoryManager::cleanup()
+MemoryManager::ThreadGuard::~ThreadGuard()
 {
-	for( MemoryPoolVector::iterator it = s_memoryPools.begin(); it != s_memoryPools.end(); ++it )
-	{
-		MemoryHelper::alignedFree( ( *it ).m_pool );
-		MemoryHelper::alignedFree( ( *it ).m_free );
+	if (--thread_guard_depth == 0) {
+		rpmalloc_thread_initialize();
```
Comment (MJansson) [inline src/core/MemoryManager.cpp]: You probably meant to do `rpmalloc_thread_finalize` here

Comment (lukas-w) [reply src/core/MemoryManager.cpp]: Absolutely. Thanks!

**Comment by gi0e5b06** [comment]
May someone try this project?
It runs at 60% CPU on my old computer.
[memory20171011.zip](https://github.com/LMMS/lmms/files/1375535/memory20171011.zip)

**Comment by qnebra** [comment]
@gi0e5b06 

On my computer with 2.2 Ghz CPU LMMS with your project opened use 14% of CPU.

**Comment by qnebra** [comment]
I am merged this PR into my local lmms and got this on cmake generate:

CMake Error at src/3rdparty/rpmalloc/CMakeLists.txt:3 (add_library):
  Cannot find source file:

    rpmalloc/rpmalloc/rpmalloc.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx


CMake Error: Cannot determine link language for target "rpmalloc".
CMake Error: CMake can not determine linker language for target: rpmalloc

**Comment by tresf** [comment]
@qnebra please see https://github.com/LMMS/lmms/pull/3873#issuecomment-335598674.  This PR switches the project to using submodules, which will take some adjustment, but will greatly reduce the amount of 3rd party source code we bundle.

**Comment by qnebra** [comment]
Thank you.
Maybe add information about submodules to a developer wiki of lmms? To avoid situations like this.

**Comment by zonkmachine** [comment]
> Maybe add information about submodules to a developer wiki of lmms?

Let's merge it first.

**Comment by qnebra** [comment]
Test & merge, of course.

**Comment by gi0e5b06** [comment]
@qnebra Thanks for testing. I'm making some tests about memory use. Actually half of the instruments are good (don't allocate extra memory) and the other half is not, mainly because of Oscillator. All oscillators are duplicated for each played note (with no new information). There must be a way to avoid that.

PS: I also rewrote BaseDetuning to avoid allocation. @softrabbit was right, a single float works fine.

**Comment by softrabbit** [comment]
>All oscillators are duplicated for each played note (with no new information). There must be a way to avoid that.

What, no new information? Frequency and phase come to mind... that being said, I'd like to see some kind of "oscillator bank" concept, where all 6 (TripleO) or 18 (Organic) Oscillators are merged into one object, preferably organizing the data and math in a way that makes it easy for compilers to optimize. (SIMD, anyone?) That might even up for some new features...

**Comment by gi0e5b06** [comment]
@softrabbit Frequency is the same than for the NotePlayHandle, same for ext_phaseOffset. PhaseOffset seems to be more or less equal to ext_phaseOffset. So only phase stays (one float), which is probably computable (I think but I may be wrong). Anyway, there is a big room for improvment here. Especially as Oscillator by many many instruments.

I wouldn't group the oscillators into a single object. Instead I would pre-allocate 2048 oscillators (for example) and reuse them when needed because all the instruments use the same Oscillator type (for now). More important, if it goes over 2048, the note should be skipped.

As a result: no memory allocation and CPU under controlled for all the instruments.

**Comment by softrabbit** [comment]
> I wouldn't group the oscillators into a single object. Instead I would pre-allocate 2048 oscillators (for example) and reuse them when needed because all the instruments use the same Oscillator type (for now). More important, if it goes over 2048, the note should be skipped.

I hate to dig up my own code as an example, but: https://github.com/LMMS/lmms/pull/2089

That's a speedup of maybe ~2x from reorganizing arrays of structs into structs of arrays. Probably even more if compiling for wider SIMD models than SSE2. That's not to say the same applies to the Oscillator situation, but it might be worth considering.

**Comment by gi0e5b06** [comment]
@softrabbit Actually I don't disagree. But that was my understanding that you talked about grouping, not restructurating. As a first step requiring little code changes, I'm suggesting to group them all in a big array (with eventually informal subgrouping in 6/18).

As for switching from "arrays of structs into structs of arrays", there are certainly possible gains. But it is a complete rewrite of the Oscillator class and of the depending instruments. And IMHO, that's something that should be only tried later. Eventually in parallel (new classes: Oscillators, Kicker2, TripleOscillator2, Organic2, etc).

**Comment by qnebra** [comment]
@gi0e5b06 

I forgot about something, on my PC your project run "as is" consume 65% of CPU. My previous tests are incorrect.

What tools can I use to test lmms perfomance? Because in my project program has low usage of system resources according to system monitoring tools, but sounds are extremely stutter and glitch.

**Comment by PhysSong** [comment]
Are there some benchmarks for cache miss rate? It is also important for performance.

**Comment by lukas-w** [comment]
@PhysSong I did a couple more benchmarks, again using `Skiessi-RandomProjectNumber14253`.

Startup and load project:

|                 |   time   |  cpu    |         Ir incl.          |   L1 Miss Sum    | Cycle Estimation |  Called
|:-------------|---------:|---------:|----------------------:|---------------------:|----------------------:|----------:|
| stable-1.2 |  2.06   | 45%    | 1.6×10⁹ <br/> (27.9%) | 3.0×10⁶ <br/> (3.79%)   | 1.7×10⁹ <br/> (23.74%)  | 6.6×10³ | 
| this PR     | 1.97    | 44%    | 471×10³ <br/> (0.01%) | 5.8×10³ <br/> (0.01%)     | 1.7×10⁴ <br/> (0.00%) | 6.6×10³ | 

Render project using CLI:

|                 |   time   |  cpu    |         Ir incl.          |   L1 Miss Sum    | Cycle Estimation |  Called
|:-------------|---------:|---------:|----------------------:|---------------------:|----------------------:|----------:|
| stable-1.2 | 07.49 | 324%  | 20×10⁹ <br/> (15.3%) | 39×10³ <br/> (7.27%) | 20×10⁹ <br/> (14.9%) | 28×10³
| this PR     | 06.30  | 359%  | 2.2×10⁶ <br/> (0.00%) | 7.2×10³ <br/> (0.00%) | 554×10³ <br/> (0.00%) | 28×10³

If KCachegrind's cycle estimation is accurate, the old `MemoryManager` is something between 36,000 (rendering) and 100,000 times (startup and loading) slower than `rpmalloc`.

Merging.

**Comment by zonkmachine** [comment]
> Maybe add information about submodules to a developer wiki of lmms?

@qnebra You are off course absolutely right here. The now updated wiki page is:
https://github.com/LMMS/lmms/wiki/Accessing-git-repository

**Comment by gi0e5b06** [comment]
My memory manager is about 35% faster than rpmalloc.
And it's not even optimized yet.
Available here:
https://github.com/gi0e5b06/lmms/blob/master/src/core/MemoryManagerArray.cpp

Drawbacks: is dedicated to LMMS, uses on average more memory

To do: auta-adaptable to the style of the user

Suggestion: make the memory manager an option configurable by the user

**Pull request merged**